sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -750,6 +804,11 @@ pi_result _pi_queue::executeCommandList(ze_command_list_handle_t ZeCommandList,
     this->ZeOpenCommandListSize = 0;
   }
 
+  bool UseCopyEngine = getZeCommandListIsCopyList(ZeCommandList);
```

Comment:
This appears very ineffective searching the entire cache of command-lists on every new submission. Can't you just grab this info directly from the fence created for this command list and passed you in `ZeFence` argument.

Reply:
To get that, we will need a level zero API call which might be expensive. One of the ways to make this faster will be to replace std::map with std::unordered_map. Is that a viable option?

Reply:
It seems like 'grabbing' this info from the fence created will require a ZE API call. I wanted to avoid that. To make this faster, we can use std::unordered_map (O(1) access) instead of std::map (O(log(n))) access time. Please advise if that is acceptable.

Reply:
I don't think there is any ZE API call that can get this information.

I think the suggestion was to eliminate the query entirely by passing the inforrmation about what "kind" of command list you have back from getAvailableCommandList, and then to pass that directly into executeCommandList, and then this would just be a reference to that parameter.  Then there would be no need to query whether a command list was a copy or compute command at all.
If you do that, then when you desire to implement batching for copy engine, then in addition to the ZeOpenCommandList that is in the _pi_queue data structure, you will also need a member that indicates whether ZeOpenCommandList is a copy or compute command list, and that will be used whenever you need to call executeOpenCommandList to pass which kind of command list there is.

Reply:
> It seems like 'grabbing' this info from the fence created will require a ZE API call.

No, why? You'd just return the `_pi_queue::command_list_fence_t` instead of the `ze_fence_handle_t` itself in the `getAvailableCommandList`, and take the info about the queue from it in the `executeCommandList`

Reply:
Please add TODO that this is going to be restructured, and the map lookup gone.

Reply:
I added a TODO before the function definition.